### ISSUE TITLE
IORING_SETUP_IOPOLL, IORING_SETUP_SQPOLL impact on TCP read event latency

### ISSUE THREAD

**Issue by lano1106** [axboe/liburing#345]
I have finally completed the 2 phases transition from epoll to io_uring of a WebSockets client application.

- First phase was simply to switch the event loop backend from epoll and use io_uring only as a polling API replacement.
- Second phase was for replacing read/write system calls with io_uring async I/O with buffer selection.

The main challenge of this second phase has been to write a custom openssl BIO module to interface openssl with io_uring.

It has been fun to do and it seems to work well without many pending problems as the client has been running for over 12 hours without any glitches.

I am looking to tweak my io_uring usage in order to minimize as much as possible my application reaction time to network events (latency).

I was wondering what io_uring devs would recommend as optimal settings to achieve this objective.

Are IORING_SETUP_IOPOLL, IORING_SETUP_SQPOLL, or both could help in improving my app latency to incoming network events?

Here are some details about my app:

- about 85 TCP connections (Handling mostly inbound traffic ~5Mbps)
- 1 UDP control server socket idle most of the time
- 1 eventfd to interrupt poll event waiting
- runs on a dedicated machine with kernel 5.12.1 compiled with nohzfull with the thread interacting with io_uring pinned on an isolated cpu at the highest realtime priority.

I have few questions about IORING_SETUP_IOPOLL:

1.
My client is configuring the TCP sockets with SO_BUSY_POLL. I have noticed that when I did switch from epoll to io_uring, my client process thread CPU usage did drop from about 30% to below 5%. I did conclude from this observation that this might be an indication that io_uring isn't honoring the SO_BUSY_POLL option. Is this correct?

Is IORING_SETUP_IOPOLL essentially doing what epoll is doing with SO_BUSY_POLL except that io_uring does it in background with a dedicated thread?

2.
Considering that I would want to busy wait poll only my TCP sockets, would that be advisable that I move my UDP socket elsewhere or is there a trivial setup such as embedding the IORING_SETUP_IOPOLL io_uring instance into another fd poll set to mix and match polled and non-polled I/O?

Is using IORING_SETUP_IOPOLL essentially assume that your fd set is sufficiently I/O intensive that it does make sense to give up the capability to wait, for a certain period, events and possibly cancel the wait, IOW let the user process spin and never wait (I need to check if eventfd is compatible with O_DIRECT but I wouldn't be surprised to find out that it isn't...)?

This code snippet from io_uring_enter() appears to support this intuition...

                /*
                 * When SETUP_IOPOLL and SETUP_SQPOLL are both enabled, user
                 * space applications don't need to do io completion events
                 * polling again, they can rely on io_sq_thread to do polling
                 * work, which can reduce cpu usage and uring_lock contention.
                 */
                if (ctx->flags & IORING_SETUP_IOPOLL &&
                    !(ctx->flags & IORING_SETUP_SQPOLL)) {
                        ret = io_iopoll_check(ctx, min_complete);
                } else {
                        ret = io_cqring_wait(ctx, min_complete, sig, argsz, ts);
                }

**Comment by lano1106** [comment]
I have started experimenting with IORING_SETUP_SQPOLL. One issue that I did encountered was that io_uring_setup was returning EINVAL.

It did turn out to be because of the value that I was trying to pass to sq_thread_cpu. Whenever I think of cpu affinity, I think about a bitmask. That is usually the case for various kernel boot parameters (ie: isolcpus, nohz_full, rcu_nocbs, irqaffinity and so on).

Therefore this is a bitmask value that I was passing to sq_thread_cpu (it must be because of the option bit name IORING_SETUP_SQ_AFF that I made the bitmask association. If it would have been IORING_SETUP_SQ_CPU, maybe I wouldn't have think that). It must be obvious for any kernel dev upon seeing the macro cpu_online(cpu) used in the io_sq_offload_create() function but it wasn't to me. I must have spent a good 20 mins getting familiar with include/linux/bitmask.h and include/linux/cpumask.h to realise that what io_uring was expecting is a cpu core id.

That being said, SQP thread is taking anything between 90-97% of cpu usage of the selected CPU. At that level of cpu usage, I cannot think of good reason for NOT using IORING_SETUP_IOPOLL!

Also, I have kinda answered myself the question that I was having.

The way IORING_SETUP_IOPOLL works, it is either the user thread calling io_uring_enter or the SQP thread that is doing the busy-wait polling. So, if you let the SQP thread doing the busy-wait polling, the user thread can call io_uring_enter() to wait for new events on IORING_SETUP_IOPOLL io_uring instance.

I need to check if an eventfd can accept the O_DIRECT flag. I expect that it doesn't but since I only use the eventfd to cancel the event waiting which is only used to terminate the process, I'll get rid of it and find another way to exit my loop...

So next update should be to tell how my IORING_SETUP_IOPOLL usage went.

**Comment by lano1106** [comment]
I have checked in fs/eventfd.c and no, this type of file doesn't support O_DIRECT flag. The only ones that it supports are O_NONBLOCK and O_CLOEXEC.

Considering that:
- I have 2 io_uring instances
- I was evaluating IORING_SETUP_IOPOLL mostly for my first io_uring instance but with my IORING_SETUP_ATTACH_WQ discovery from #349, the second instance might take advantage of instance 1 SQPOLL thread as well for "free".
- However, using IORING_SETUP_ATTACH_WQ might bring the requirement of IORING_SETUP_IOPOLL on instance 2 as well (not 100% sure about this. I'm currently seeking validation but it would make sense that it is true).
- While it could make sense to make instance 2 a busy-wait polling instance, to the contrary of instance 1, eventfd plays a central role in instance 2 setup. I simply cannot toss instance 2 eventfd away like I have imagined doing with the instance 1 eventfd.

So I see 2 options:
1. I study how IORING_SETUP_IOPOLL works and figure out a way on how eventfd could work with it.
2. I find an existing substitute file type that is compatible with IORING_SETUP_IOPOLL that could be used to interrupt the completion waiting. (pipe, unix socket, AF_INET socket, anything). It would work but that would be kinda silly when a the lightweight eventfd could be modified to do it...

**Comment by lano1106** [comment]
I am in the process of sorting out various concepts in my mind...

IORING_SETUP_IOPOLL is using the struct file operation iopoll which seems to not be widely supported. I am not even sure that io_uring does check if the file operation is present before onboarding a fd in an iopoll instance.

As far as can tell very few types of files implement the iopoll operation. The only one that I have found are:
- ext4
- xfs
- gfs2
- and block device files

from fs/fcntl.c:
an O_DIRECT file is one that expose filp->f_mapping->a_ops->direct_IO
is there a 1-to-1 correlation between exposing iopoll and direct_IO? IDK

but just greping '\.direct_IO' in the kernel source code tree, there seems to have much more file types exposing direct_IO than there are exposing iopoll.

So in the best-case scenario, I think one way to improve the io_uring_setup man page would be to clarify this point concerning IORING_SETUP_IOPOLL 

Bottom line, I naively figured that TCP sockets could be used with IOPOLL because busy-wait poll is supported for them (via napi_busy_loop() and driven with the option SO_BUSY_POLL) and that epoll is leveraging this feature and I saw TCP and O_DIRECT in the same blog post:
https://blog.cloudflare.com/io_submit-the-epoll-alternative-youve-never-heard-about/

So, busy-wait polling on sockets doesn't seem to be supported yet but this is definitely achievable since epoll does it.
I'll try to see if I could be able to modify io_uring to do that... I'll offer a patch if I succeed making something that works...

**Comment by lano1106** [comment]
now that I have read the source code and I have learned about iopoll file operation, the man page does specify this... I'm not sure that it is clear enough to the common user that I am...

**Comment by lano1106** [comment]
I did stumble into SO_PREFER_BUSY_POLL option introduced in 5.11... I might get away with my busy-polling without any changes after all...

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7fd3253a7de6a317a0683f83739479fb880bffc8

**Comment by lano1106** [comment]
> IORING_SETUP_IOPOLL is using the struct file operation iopoll which seems to not be widely supported. I am not even sure that io_uring does check if the file operation is present before onboarding a fd in an iopoll instance.
```
	if (ctx->flags & IORING_SETUP_IOPOLL) {
		if (!(kiocb->ki_flags & IOCB_DIRECT) ||
		    !kiocb->ki_filp->f_op->iopoll)
			return -EOPNOTSUPP;
```

**Comment by beef9999** [comment]
> Are IORING_SETUP_IOPOLL, IORING_SETUP_SQPOLL, or both could help in improving my app latency to incoming network events?

I'm asking the same question. What is your conclusion then?

**Comment by lano1106** [comment]
IORING_SETUP_IOPOLL is meant only for File Systems. O_DIRECT is meaningless outside file systems. This is meant to bypass the kernel page cache and go directly into the disk device.

For TCP sockets, what you want is to do is use NAPI busy loop.

I wrote a patch to add NAPI busy loop from io_uring:
https://lore.kernel.org/io-uring/cover.1646777484.git.olivier@trillion01.com/T/#u

**Comment by JonKohler** [comment]
That looks like a pretty slick patch, thanks for adding that!

**Comment by ileixe** [comment]
Hi, I found this issue while I'm looking for the way to introduce new character device with io_uring supported. As I'm not very familiar with kernel APIs, I'm stuggling with understanding what is required for `IOURING_SETUP_IOPOLL`. Currently I'm stuck with the same location what you described below as we're using character device which seems not to be valid with `O_DIRECT`

> IORING_SETUP_IOPOLL is using the struct file operation iopoll which seems to not be widely supported. I am not even sure that io_uring does check if the file operation is present before onboarding a fd in an iopoll instance.

I saw your patch but still did not catch how it bypass the `O_DIRECT` validation. Does your 'napi code path' entirely bypass the `io_submit_sqe`?

**Comment by axboe** [comment]
@ileixe please don't just take over some older issue, file a discussion or new issue. As you may note, this one is closed and done.
